src/librustc/ty/sty.rs

```diff
@@ -556,9 +556,39 @@ pub struct ProjectionTy<'tcx> {
     /// The trait reference `T as Trait<..>`.
     pub trait_ref: ty::TraitRef<'tcx>,
 
-    /// The name `N` of the associated type.
-    pub item_name: Name,
+    /// The DefId of the the associated type `N`.
+    ///
+    /// Note that this is not the DefId of the TraitRef containing this
+    /// associated type, which is in tcx.associated_item(item_def_id).container.
+    pub item_def_id: DefId,
 }
+
+impl<'a, 'tcx> ProjectionTy<'tcx> {
+    pub fn from_ref_and_name(
+        tcx: TyCtxt,
+        trait_ref: ty::TraitRef<'tcx>,
+        item_name: Name) -> ProjectionTy<'tcx> {
+        // FIXME(tschottdorf): elsewhere in the code, we don't compare on item_name,
+        // but on an ident obtained as such:
+        //
+        // let ident = self.tcx.adjust(item_name, def_id, self.body_id).0;
+        //
+        // (There is no body_id in scope here). Is the below incorrect?
```

Comment:
Yeah, bother, I don't know how hygiene and "associated item lookup" are supposed to interact here. @jseyfried may have thoughts.